### PR TITLE
add new colloector

### PR DESCRIPTION
**Pull request by haiyuanhe**
spark and storm collector

### CODE REVIEW

File: collectors/conf/storm.conf [hunk 1]
```diff
@@ -0,0 +1,6 @@
+[base]
+enabled: false
+interval: 120
```
Comment (wangy1931) [inline collectors/conf/storm.conf]: why 120 seconds, is each collecting too expensive? I feel 120 seconds is too long.

File: collectors/conf/spark.conf [hunk 2]
```diff
@@ -0,0 +1,31 @@
+[base]
+enabled: True
+interval: 120
+
```
Comment (wangy1931) [inline collectors/conf/spark.conf:4]: why 120 seconds, is each collecting too expensive? I feel 120 seconds is too long.

File: collectors/builtin/spark.py [hunk 3]
```diff
@@ -0,0 +1,373 @@
...
+    def __call__(self):
+        spark_apps = self._get_running_apps()
+        self._spark_job_metrics(spark_apps)
+        self._spark_stage_metrics(spark_apps)
+        self._spark_executor_metrics(spark_apps)
+
+    def _get_running_apps(self):
+        if self.spark_cluster_mode is None:
+            self.log_exception('spark_cluster_node must not be null')
+            return None
```
Comment (wangy1931) [inline collectors/builtin/spark.py]: how to handle None, should you fail fast

File: collectors/builtin/spark.py [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+        self.port = self.get_config('spark_port', 8080)
+        self.spark_url = 'http://%s:%s' % (self.host, self.port)
+
+    def __call__(self):
+        spark_apps = self._get_running_apps()
+        self._spark_job_metrics(spark_apps)
+        self._spark_stage_metrics(spark_apps)
+        self._spark_executor_metrics(spark_apps)
+
+    def _get_running_apps(self):
```
Comment (wangy1931) [inline collectors/builtin/spark.py:170]: _get_running_apps is an expensive operations, we probably shouldn't run in each __call__

File: collectors/conf/storm.conf (hunk 1 above)
Comment (haiyuanhe) [reply collectors/conf/storm.conf]: Just for test,  I don't want to print so fast at the terminal. And collecting in 60 second would be great.

File: collectors/builtin/spark.py (hunk 4 above)
Comment (haiyuanhe) [reply collectors/builtin/spark.py:170]: if I cache the result . I should choose a time range to refresh the job lists. and the new job status would be missing

File: collectors/builtin/spark.py (hunk 3 above)
Comment (haiyuanhe) [reply collectors/builtin/spark.py]: yes, if return none and fail fast .
And  I have some question while I was writing the collector .
After our customer deploy collector with a section script , and we haven't  git any configure interface out of them . Do they need to open config file and overwrite the property?

**Comment by haiyuanhe** [comment]
@wangy1931

File: collectors/builtin/spark.py [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+            self.log_exception("Can not get anything form SPARK_MASTER_APP_PATH %s" %
+                               url_join_3(self.spark_url,
+                                          REST_API['SPARK_MASTER_APP_PATH'],
+                                          "appId=" + app_id))
+            raise HTTPError(app_page)
+
+        # parser html and get app_detail_direct_link
+        parser = SparkParser()
+        parser.feed(app_page.text)
+        parser.close()
```
Comment (wangy1931) [inline collectors/builtin/spark.py]: check whether close needs to be in finally clause

File: collectors/builtin/spark.py (hunk 4 above)
Comment (wangy1931) [reply collectors/builtin/spark.py:170]: ok. looks good then.

File: collectors/builtin/spark.py (hunk 3 above)
Comment (wangy1931) [reply collectors/builtin/spark.py]: yes, currently we have to manually config. Hopefully default value will serve most cases.

File: collectors/builtin/storm.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    def __init__(self, config, logger, readq):
+        super(Storm, self).__init__(config, logger, readq)
+        
+        self.port = self.get_config('port', 8080)
+        self.host = self.get_config('host', "localhost")
+        self.http_prefix = 'http://%s:%s' % (self.host, self.port)
+
+    def __call__(self):
+        try:
+            self._cluster_loader()
```
Comment (wangy1931) [inline collectors/builtin/storm.py:82]: suggest try... catch for each function call, so one failure will not affect others.

**Review by wangy1931 — APPROVED** [review]

**Pull request merged**